solidity/src/CadenceRandomConsumer.sol

```diff
@@ -2,30 +2,84 @@
  * revert on undesirable random results.
  */
 abstract contract CadenceRandomConsumer is CadenceArchWrapper {
+    using Xorshift128plus for Xorshift128plus.PRG;
+
     // A struct to store the request details
     struct Request {
         // The Flow block height at which the request was made
         uint64 flowHeight;
         // The EVM block height at which the request was made
         uint256 evmHeight;
+        // Whether the request has been fulfilled
+        bool fulfilled;
     }
 
     // Events
-    event RandomnessRequested(uint256 requestId, uint64 flowHeight, uint256 evmHeight);
-    event RandomnessFulfilled(uint256 requestId, uint64 flowHeight, uint256 evmHeight, uint256 randomResult);
+    event RandomnessRequested(uint256 indexed requestId, uint64 flowHeight, uint256 evmHeight);
+    event RandomnessSourced(uint256 indexed requestId, uint64 flowHeight, uint256 evmHeight, bytes32 randomSource);
+    event RandomnessFulfilled(uint256 indexed requestId, uint64 randomResult);
 
     // A list of requests where each request is identified by its index in the array
     Request[] private _requests;
     // A counter to keep track of the number of requests made, serving as the request ID
     uint256 private _requestCounter;
 
+    ///////////////////
+    // PUBLIC FUNCTIONS
+    ///////////////////
+
+    /**
+     * @dev This method checks if a request can be fulfilled.
+     *
+     * @param requestId The ID of the randomness request to check.
+     * @return canFulfill Whether the request can be fulfilled.
+     */
+    function canFulfillRequest(uint256 requestId) public view returns (bool) {
+        uint256 requestIndex = requestId - 1;
+        if (requestIndex >= _requests.length) {
+            return false;
+        }
+        Request storage request = _requests[requestIndex];
+        uint64 flowHeight = _flowBlockHeight();
+        return !request.fulfilled && request.flowHeight < flowHeight;
+    }
+
+    /////////////////////
+    // INTERNAL FUNCTIONS
+    /////////////////////
+
+    /**
+     * @dev This method returns a ***REVERTIBLE** random number in the range [min, max].
+     * NOTE: The fact that this method is revertible means that it should only be used in cases where the initiating
+     * caller is trusted not to revert on the result.
+     *
+     * @param min The minimum value of the range (inclusive).
+     * @param max The maximum value of the range (inclusive).
+     * @return random The random number in the range [min, max].
+     */
+    function _getRevertibleRandomInRange(uint64 min, uint64 max) internal view returns (uint64) {
+        bytes memory seed = abi.encodePacked(_aggregateRevertibleRandom256());
+        bytes memory salt = abi.encodePacked(block.number);
+
+        // Instantiate a PRG with the aggregate bytes and salt with current block number
+        Xorshift128plus.PRG memory prg;
+        prg.seed(seed, salt);
+
+        return _getNumberInRange(prg, min, max);
```

Comment:
> we can loop and draw 64 bits from revertibleRandom each time, applying the mask and checking against range.

As currently implemented, this method is used for both revertible and non-revertible random numbers within ranges. The seed for the PRGs differ - aggregate bytes for revertible random and random source for non-revertible  - but this method supports fulfilling both sorts of requests.

My intuition is that sourcing additional bytes from `revertibleRandom` would then mean that non-revertible results are no longer deterministic and I could revert the "reveal" step indefinitely until I reveal a random result that matches my commitment. This is based on my assumption that two PRGs with the same seed and salt result in identical paths. Is that accurate?

Reply:
Yes I've seen that you opted for a common code for both revertible and non-revertible randomness (the PRG part seeded by 256 bits + the iterations till we get a valid random), which is understandable. 
I was only commenting that if only revertible randomness is needed, we could omit the PRG part entirely, but no need to do in this repo. 

> My intuition is that sourcing additional bytes from revertibleRandom would then mean that non-revertible results are no longer deterministic

Not sure I followed correctly but you're right if we aim for non-revertible randomness, the function `revertibleRandom` shouldn't be used at all. 
and yes a PRG with the same inputs always give the same sequences.

> I used the block number just for consistency, but could I just use 0 as the salt for the revertible random PRG?
for revertible randomness, either way is fine. I'm just commenting that `0` (or any constant) works in this case and that the block number wouldn't add extra diversification, because the internal PRG (in the FVM) is already diversified for the specific block and transaction.